File: 343-integer-break.go
func integerBreak(n int) int {
    if n == 2{
        return 1
    }
    if n == 3{
        return 2
    }
    if n == 4{
        return 4
    }
    base := 1
    for n > 4 {
        base = base * 3
        n -= 3
    }
    base *= n
    return base
}

File: bwu/dp/343-integer-break.py
class Solution(object):
    def integerBreak(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n < 2:
            return 0
        
        dp = [1] * (n + 1)
        for i in xrange(3, n + 1):
            for j in xrange(1, i):
                dp[i] = max(dp[i], max(j, dp[j]) * max((i -j), dp[i - j]))
                
        return dp[n]
